Simplify the OpInfos for norm / linalg_norm

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #76551
* #76547
* __->__ #76545

This will be helpful later on when we add tests for `dtype` (after correcting its gradient formula).

lezcano committed:
Simplify the OpInfos for norm / linalg_norm

This will be helpful later on if we want to start testing dtype (for
which the gradient formula is currently wrong).

[ghstack-poisoned]

Lezcano commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1842,6 +1842,9 @@ def sample_inputs_norm(op_info, device, dtype, requires_grad, **kwargs):
         ((S, S), (-2,), 'neg_2'),
         ((S, S), (-0.5,), 'neg_0_5'),
         ((S, S), (-1.5,), 'neg_1_5'),
+        # FIXME gradgrad and noncotiguous_samples fails for these on CPU, how about that
+        # ((S, S), (inf,), 'inf'),
+        # ((S, S), (-inf,), 'neg_inf'),
```
This was an interesting find, and it seems to be related to the odd fix `linalg.vector_norm` has. See https://github.com/pytorch/pytorch/pull/76547#discussion_r861201122.

mruberry replied:
You can conditionally include these when CUDA samples are requested

lezcano committed:
Update on "Simplify the OpInfos for norm / linalg_norm"


This will be helpful later on if we want to start testing dtype (for
which the gradient formula is currently wrong).

[ghstack-poisoned]

ngimel commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1932,63 +1930,14 @@ def sample_inputs_norm_inf(op_info, device, dtype, requires_grad, **kwargs):
+    ords = (inf, 2, 1, 0, 0.9, -2.1, -inf)
 
-    return inputs
+    for size, ord_, dim, keepdim in product(sizes, ords, dims, (True, False)):
```
you are no longer testing tuple of dims for 2d inputs?

Lezcano replied:
I had that in mind, and then I forgot to add it. Thank you for the catch!

lezcano committed:
Update on "Simplify the OpInfos for norm / linalg_norm"


This will be helpful later on when we add tests for `dtype` (after correcting its gradient formula).

[ghstack-poisoned]

Lezcano commented:
This is ready for another review and perhaps a merge @ngimel @mruberry

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1910,7 +1910,13 @@ def sample_inputs_norm(op_info, device, dtype, requires_grad, **kwargs):
-    )
+    ]
+
+    # FIXME gradgrad and noncotiguous_samples fail on inf and -inf norms on CPU because of vectorization
```
Nice note and device type check

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1937,15 +1943,10 @@ def sample_inputs_norm(op_info, device, dtype, requires_grad, **kwargs):
-        new_args = copy.deepcopy(list(args))
-        new_args[1] *= -1
-        cases_negdim.append((shape, tuple(new_args), name.replace("_dim", "_neg_dim")))
+    cases_negdim = ((shape, args[:1] + (-args[1],) + args[2:], name.replace("_dim", "_neg_dim"))
```
Comment here please

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1992,63 +1993,19 @@ def sample_inputs_norm_inf(op_info, device, dtype, requires_grad, **kwargs):
-                    keepdim=keepdim,
-                    dim=dim)))
+    for size, ord_, keepdim in product(sizes, ords, (True, False)):
+        for dim in dims:
```
Why a separate iteration for dims here?

Lezcano replied:
To be able to add the extra dim case of `(-1, 0)` when `len(size) >= 2`.

mruberry replied:
Oh, right, sorry it just appeared so separated on GitHub I failed to make the connection

mruberry reviewed (approved):
Cool! LGTM

lezcano committed:
Update on "Simplify the OpInfos for norm / linalg_norm"


This will be helpful later on when we add tests for `dtype` (after correcting its gradient formula).

[ghstack-poisoned]

lezcano committed:
Update on "Simplify the OpInfos for norm / linalg_norm"


This will be helpful later on when we add tests for `dtype` (after correcting its gradient formula).

[ghstack-poisoned]

lezcano committed:
Update on "Simplify the OpInfos for norm / linalg_norm"


This will be helpful later on when we add tests for `dtype` (after correcting its gradient formula).

[ghstack-poisoned]

lezcano committed:
Update on "Simplify the OpInfos for norm / linalg_norm"


This will be helpful later on when we add tests for `dtype` (after correcting its gradient formula).

[ghstack-poisoned]

lezcano committed:
Update on "Simplify the OpInfos for norm / linalg_norm"


This will be helpful later on when we add tests for `dtype` (after correcting its gradient formula).

[ghstack-poisoned]

ngimel commented:
@pytorchbot merge this
